pkg/render/applicationlayer.go

```diff
@@ -0,0 +1,308 @@
+		},
+	}
+
+	ds := &appsv1.DaemonSet{
+		TypeMeta: metav1.TypeMeta{Kind: "DaemonSet", APIVersion: "apps/v1"},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      L7LogCollectorDeamonsetName,
+			Namespace: CalicoSystemNamespace,
+		},
+		Spec: appsv1.DaemonSetSpec{
+			Selector: &metav1.LabelSelector{MatchLabels: map[string]string{"k8s-app": L7LogCollectorDeamonsetName}},
+			Template: *podTemplate,
+			UpdateStrategy: appsv1.DaemonSetUpdateStrategy{
+				RollingUpdate: &appsv1.RollingUpdateDaemonSet{
+					MaxUnavailable: &maxUnavailable,
+				},
+			},
+		},
+	}
+
+	return ds
+}
+
+func (c *ApplicationLayerComponent) containers() []corev1.Container {
+
+	var containers []corev1.Container
+
+	proxy := corev1.Container{
+		Name:  ProxyContainerName,
+		Image: c.proxyImage,
+		Command: []string{
+			"envoy", "-c", "/etc/envoy/envoy-config.yaml",
+		},
+		SecurityContext: &corev1.SecurityContext{
+			Privileged: ptr.BoolToPtr(false),
+			Capabilities: &corev1.Capabilities{
+				Add: []corev1.Capability{"NET_ADMIN", "NET_RAW"},
+			},
+			RunAsUser:  ptr.Int64ToPtr(0),
+			RunAsGroup: ptr.Int64ToPtr(0),
+		},
+		Env:          c.proxyEnv(),
+		VolumeMounts: c.proxyVolMounts(),
+	}
+	containers = append(containers, proxy)
+
+	collector := corev1.Container{
+		Name:         L7CollectorContainerName,
+		Image:        c.collectorImage,
+		Env:          c.collectorEnv(),
+		VolumeMounts: c.collectorVolMounts(),
+	}
+	containers = append(containers, collector)
+
+	return containers
+}
+
+func (c *ApplicationLayerComponent) proxyEnv() []corev1.EnvVar {
+	envs := []corev1.EnvVar{
+		{Name: "ENVOY_UID", Value: "0"},
+		{Name: "ENVOY_GID", Value: "0"},
+	}
+
+	return envs
+}
+
+func (c *ApplicationLayerComponent) collectorEnv() []corev1.EnvVar {
+	envs := []corev1.EnvVar{
+		{Name: "LOG_LEVEL", Value: "Info"},
+		{Name: "FELIX_DIAL_TARGET", Value: "/var/run/felix/nodeagent/socket"},
+	}
+	// set rate limiters if provided
+	if c.l7Spec.LogRequestsPerInterval != nil {
+		envs = append(envs, corev1.EnvVar{
+			Name:  "ENVOY_LOG_REQUESTS_PER_INTERVAL",
+			Value: strconv.FormatInt(*c.l7Spec.LogRequestsPerInterval, 10),
+		})
+	}
+
+	if c.l7Spec.LogIntervalSeconds != nil {
+		envs = append(envs, corev1.EnvVar{
+			Name:  "ENVOY_LOG_INTERVAL_SECONDS",
+			Value: strconv.FormatInt(*c.l7Spec.LogIntervalSeconds, 10),
+		})
+	}
+
+	return envs
+}
+
+// tolerations creates the node's toleration.
+func (c *ApplicationLayerComponent) tolerations() []corev1.Toleration {
+	// ensures that l7 log collector pods are scheduled on master node as well
+	toleration := []corev1.Toleration{
+		{Key: "node-role.kubernetes.io/master", Operator: corev1.TolerationOpExists, Effect: corev1.TaintEffectNoSchedule},
+	}
+
+	return toleration
+}
+
+func (c *ApplicationLayerComponent) volumes() []corev1.Volume {
+
+	var volumes []corev1.Volume
+
+	volumes = append(volumes, corev1.Volume{
+		Name: EnvoyLogsKey,
+		VolumeSource: corev1.VolumeSource{
+			EmptyDir: &corev1.EmptyDirVolumeSource{},
```

Comment:
What is this? adding a volume that is empty? what's the point?
If this is needed please include a comment why so it is clear why we need it.

Reply:
added a comment explaining the reason d17bee1